lib/versionbot.ts

```diff
@@ -0,0 +1,479 @@
+
+        // Only for opened or synced actions.
+        if ((data.action !== 'opened') && (data.action !== 'synchronize')) {
+            return Promise.resolve();
+        }
+
+        return this.gitCall(githubApi.pullRequests.getCommits, {
+            owner: owner,
+            repo: name,
+            number: pr.number
+        }).then((commits: GithubBot.Commit[]) => {
+            let changetypeFound: boolean = false;
+            // Go through all the commits. We're looking for, at a minimum, a 'change-type:' tag.
+            for (let index = 0; index < commits.length; index += 1) {
+                const commit: GithubBot.Commit = commits[index];
+                const commitMessage: string = commit.commit.message;
+                const invalidCommit = !commitMessage.match(/^change-type:\s*(patch|minor|major)\s*$/mi);
+
+                if (!invalidCommit) {
+                    changetypeFound = true;
+                    break;
+                }
+            }
+
+            // If we found a change-type message, then mark this commit as ok.
+            if (changetypeFound) {
+                return this.gitCall(githubApi.repos.createStatus, {
+                    owner: owner,
+                    repo: name,
+                    sha: head.sha,
+                    state: 'success',
+                    description: 'Found a valid Versionist `Change-Type` tag',
+                    context: 'Versionist'
+                }).return(true);
+            }
+
+            // Else we mark it as having failed.
+            this.log(ProcBot.LogLevel.DEBUG, `${action.name}: No valid 'Change-Type' tag found, failing last commit`);
+            return this.gitCall(githubApi.repos.createStatus, {
+                owner: owner,
+                repo: name,
+                sha: head.sha,
+                state: 'failure',
+                description: 'None of the commits in the PR have a `Change-Type` tag',
+                context: 'Versionist'
+            }).return(false);
+        }).then(() => {
+            // If the author was Versionbot and the file marked was CHANGELOG, then
+            // we are now going to go ahead and perform a merge.
+            //
+            // Get the list of commits for the PR, then get the very last commit SHA.
+            return this.gitCall(githubApi.repos.getCommit, {
+                owner,
+                repo: name,
+                sha: head.sha
+            }).then((headCommit: GithubBot.Commit) => {
+                // We will go ahead and perform a merge if we see Versionbot has committed
+                // something with 'CHANGELOG.md' in it.
+                const commit = headCommit.commit;
+                const files = headCommit.files;
+
+                if ((commit.committer.name === process.env.VERSIONBOT_NAME) &&
+                    _.find(files, (file: GithubBot.CommitFile) => {
+                        return file.filename === 'CHANGELOG.md';
+                    })) {
```

Comment:
The indentation here plus the double indentation of the block below and the callback inline in the test makes this really really confusing to read, for me at least.